Source/WebKit/UIProcess/Notifications/glib/NotificationService.cpp

```diff
@@ -0,0 +1,310 @@
+            m_capabilities.add(Capabilities::BodyImages);
+        else if (!g_strcmp0(capability, "body-markup"))
+            m_capabilities.add(Capabilities::BodyMarkup);
+        else if (!g_strcmp0(capability, "icon-multi"))
+            m_capabilities.add(Capabilities::IconMulti);
+        else if (!g_strcmp0(capability, "icon-static"))
+            m_capabilities.add(Capabilities::IconStatic);
+        else if (!g_strcmp0(capability, "persistence"))
+            m_capabilities.add(Capabilities::Persistence);
+        else if (!g_strcmp0(capability, "sound"))
+            m_capabilities.add(Capabilities::Sound);
+    }
+}
+
+bool NotificationService::showNotification(const WebNotification& notification)
+{
+    if (!m_proxy)
+        return false;
+
+    auto findNotificationByTag = [&](const String& tag) -> Notification {
+        if (tag.isEmpty())
+            return Notification();
+
+        uint64_t notificationID = 0;
+        for (const auto& it : m_notifications) {
+            if (it.value.tag == tag) {
+                notificationID = it.key;
+                break;
+            }
+        }
+
+        return notificationID ? m_notifications.take(notificationID) : Notification({ 0, { }, tag });
+    };
+
+    auto addResult = m_notifications.add(notification.notificationID(), findNotificationByTag(notification.tag()));
+
+    if (shouldUsePortal()) {
+        GVariantBuilder builder;
+        g_variant_builder_init(&builder, G_VARIANT_TYPE_VARDICT);
+
+        g_variant_builder_add(&builder, "{sv}", "title", g_variant_new_string(notification.title().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "body", g_variant_new_string(notification.body().utf8().data()));
+        g_variant_builder_add(&builder, "{sv}", "default-action", g_variant_new_string("default"));
+        addResult.iterator->value.portalID = createVersion4UUIDString();
+        g_dbus_proxy_call(m_proxy.get(), "AddNotification", g_variant_new("(s@a{sv})", addResult.iterator->value.portalID.utf8().data(), g_variant_builder_end(&builder)),
+            G_DBUS_CALL_FLAGS_NONE, -1, nullptr, nullptr, nullptr);
+    } else {
+        GVariantBuilder actionsBuilder;
+        g_variant_builder_init(&actionsBuilder, G_VARIANT_TYPE("as"));
+
+        if (m_capabilities.contains(Capabilities::Actions)) {
+            g_variant_builder_add(&actionsBuilder, "s", "default");
+            g_variant_builder_add(&actionsBuilder, "s", _("Acknowledge"));
+        }
+
+        const char* applicationID = nullptr;
+        if (auto* app = g_application_get_default())
+            applicationID = g_application_get_application_id(app);
+
+        GVariantBuilder hintsBuilder;
+        g_variant_builder_init(&hintsBuilder, G_VARIANT_TYPE("a{sv}"));
+        if (applicationID)
+            g_variant_builder_add(&hintsBuilder, "{sv}", "desktop-entry", g_variant_new_string(applicationID));
+
+        auto* value = static_cast<GValue*>(fastZeroedMalloc(sizeof(GValue)));
```

Comment:
Can you use `UniqueRef` for this?

If not, you can use a `std::unique_ptr` with a custom deleter and `WTFMove` it into the callback.

Reply:
I'll try